jina/drivers/__init__.py

```diff
@@ -242,9 +244,11 @@ def pre_traverse(_docs, traverse_on, context_doc=None):
                     self._apply_all(_docs, context_doc, traverse_on, *args, **kwargs)
 
                 for d in _docs:
+                    # check if apply on the current level
                     if self.is_apply and d.level_depth >= self._depth_start:
                         self._apply(d, context_doc, traverse_on, *args, **kwargs)
-                    if d.level_depth < self._depth_end:
+                    # check if apply to the next level
+                    if (d.level_depth + 1) < self._depth_end:
```

Comment:
what if we are traversing on matches? next level is not ensured right?

Reply:
> `(d.level_depth + 1)`

+1, -1, is a pretty bad way to handle boundary. Can you explain why do you need this change?

Reply:
if you look at the changes of this PR, it seems that you `(d.level_depth + 1)` but also changes all right value `+1`, i.e. from `[1, 1]` to `[1,2]`. **Note, these two changes cancel each other out**, it means there must be something else causing your bug.

Reply:
the following test will fail because the `_apply` will run for `self._depth_end` in the case of `pre_traverse`

```
FAILED tests/unit/drivers/test_reduce_all_driver.py::ReduceAllDriverTestCase::test_merge_chunks_with_different_modality
FAILED tests/unit/drivers/querylang/test_querylang_drivers.py::QueryLangTestCase::test_filter_ql_modality
FAILED tests/unit/flow/test_flow.py::FlowTestCase::test_flow_with_modalitys_simple
```

Reply:
> what if we are traversing on matches? next level is not ensured right?

Nope, the following codes tested.
```
In [7]: d = jina_pb2.Document()
In [8]: getattr(d, 'matches')
Out[8]: []
In [9]: m = d.matches.add()
In [10]: getattr(m, 'matches')
Out[10]: []
```

Reply:
okay noted, then let's keep it in this way. `pre_traverse` and `post_traverse` work differently that's fine.

Reply:
Well, this is expected, but what if `doc` actually has `matches` and we are traversing on them for this level? We are skipping under the assumption that next iteration will be in a different level right?

Reply:
> > what if we are traversing on matches? next level is not ensured right?
> 
> Nope, the following codes tested.
> 
> ```
> In [7]: d = jina_pb2.Document()
> In [8]: getattr(d, 'matches')
> Out[8]: []
> In [9]: m = d.matches.add()
> In [10]: getattr(m, 'matches')
> Out[10]: []
> ```

'd' and `m` have no matches when the getattr is called

Reply:
> > `(d.level_depth + 1)`
> 
> +1, -1, is a pretty bad way to handle boundary. Can you explain why do you need this change?

Good point. I've made another try.

Reply:
> Well, this is expected, but what if `doc` actually has `matches` and we are traversing on them for this level? We are skipping under the assumption that next iteration will be in a different level right?

@JoanFM I'm not sure I've got your point. Note that all the `matches` of the `doc` should have the same `level_depth` as the `doc`. Is it related to your concern?

```
doc: level_depth=k
  |-- matches: level_depth=k
  |        |-- matches: level_depth=k
  |        |-- matches: level_depth=k
  |                |-- matches: level_depth=k
  |                |-- matches: level_depth=k
  |
  |-- matches: level_depth=k
```

Reply:
that's why I do not see while we have to avoid doing `pre_traverse` checking a level that may not apply. But anyway I am sure you had a proper analysis of the issue.